Add Authorization Endpoint and More

dshanske opened the pull request.

dshanske committed:
Add Login by URL option and adjust authenticate option to accommodate

dshanske committed:
Fixes for Token Endpoint Revokation and Start of Authorization Endpoint

dshanske committed:
Initial commit for authorization endpoint

dshanske committed:
Return user url as me property instead of originally passed URL

aaronpk commented:
This is missing the default value set to the new built-in authorization endpoint at https://github.com/dshanske/wordpress-indieauth/blob/master/indieauth.php#L64

aaronpk commented:
Is [this](https://github.com/dshanske/wordpress-indieauth/blob/token/includes/class-indieauth-token-endpoint.php#L177) supposed to be `if ( $return )` instead of `$token`? In my case the set token method returned false but the API call still gave the client a token even though it was not stored in the database.

aaronpk commented:
* I have my test site set to "single-user mode" in the IndieWeb plugin. 
* There is no value written to the `user_url` column for my user.

When the `get_user_by_identifier` function runs, it gets all the way to the [query by login function](https://github.com/dshanske/wordpress-indieauth/blob/token/includes/functions.php#L149) and no results are returned.

Maybe this plugin should also check if there is only one user in wordpress and return that user as a fallback?

dshanske commented:
As discussed, the idea of using the Indieweb plugin's single_author_setting should do the trick there.

aaronpk commented:
Something is going wrong with the query string parameters between all the redirect steps.

after i click log in on my client:

`https://tiny.xyz.dev/wp-login.php?redirect_to=https%3A%2F%2Ftiny.xyz.dev%2Fwp-login.php%3Fredirect_to%3Dhttps%253A%252F%252Fmicropub.rocks.dev%252Fendpoints%252Fcallback%26reauth%3D1%26action%3Dindieauth%26client_id%3Dhttps%3A%2F%2Fmicropub.rocks.dev%2F%26state%3D7042033%26scope%3Dcreate+update+delete+undelete%26me%3Dhttps%3A%2F%2Ftiny.xyz.dev`

note that it has the correct `+` between the scope values here.

after i type my password, I get taken to this URL:

`https://tiny.xyz.dev/wp-login.php?redirect_to=https%3A%2F%2Fmicropub.rocks.dev%2Fendpoints%2Fcallback&reauth=1&action=indieauth&client_id=https://micropub.rocks.dev/&state=7042033&scope=createupdatedeleteundelete&me=https://tiny.xyz.dev`

all the spaces are gone from the scope parameter.

aaronpk commented on templates/indieauth-authorize-form.php:
```diff
@@ -0,0 +1,53 @@
+<?php echo get_avatar( $current_user->ID, '78' ); ?>
+<?php
+	printf(
+		'<p>' . __( 'The app <strong>%1$s</strong> would like to access your site, <strong>%2$s</strong> using the credentials of <strong>%3$s</strong>(%4$s).', 'indieauth' ) . '</p>',
```
There should be an extra space between `</strong>(`.

Also it would be nice to skip showing the `(%4$s)` part if that is the same value as `%3`, for example when the user has not set a display name so both strings are the username.

aaronpk commented:
The authorization screen could use a bit of UI cleanup.

![screenshot 2018-04-04 11 57 53](https://user-images.githubusercontent.com/113001/38328322-73be62ae-37ff-11e8-9db3-3b3c313dd771.png)

with a little more spacing around things it could look like this:

![screenshot 2018-04-04 12 04 11](https://user-images.githubusercontent.com/113001/38328766-9575c90e-3800-11e8-8fd8-53c7b480fcb7.png)

aaronpk committed:
visual updates for the authorization form

this makes it look more like the example screenshot in https://github.com/indieweb/wordpress-indieauth/pull/20#issuecomment-378711285

aaronpk commented:
I've broken these out into new issues so they'll be easier to discuss:

* https://github.com/indieweb/wordpress-indieauth/issues/24
* https://github.com/indieweb/wordpress-indieauth/issues/23
* https://github.com/indieweb/wordpress-indieauth/issues/22
* https://github.com/indieweb/wordpress-indieauth/issues/21

aaronpk commented on includes/class-indieauth-authorization-endpoint.php:
```diff
@@ -0,0 +1,148 @@
+				'action'    => 'indieauth',
+				'client_id' => $params['client_id'],
+				'state'     => $params['state'],
+				'scope'     => isset( $params['scope'] ) ? $params['scope'] : 'create update',
```
don't set the default `create update` scope if `response_type=id`

aaronpk reviewed (changes requested):
I've collected a bunch of issues in a milestone so they can be discussed easier: https://github.com/indieweb/wordpress-indieauth/milestone/1

dshanske committed:
Check for redirects on same host

dshanske committed:
Merge pull request #2 from aaronpk/token

visual updates for the authorization form

dshanske committed:
Add default to user search using Indieweb settings

dshanske committed:
Protect against wordpress redirect stripping by urlencoding scope

dshanske committed:
Do not ask for credentials if already logged in

dshanske committed:
Add alternate template for authorization vs authentication

dshanske committed:
Update README

dshanske committed:
Do not pass through scope if request is for code

dshanske committed:
By default do not enable checkbox on login form.

dshanske committed:
Add Hooks to Allow Adding to Forms

dshanske committed:
Do not generate code until authorized

dshanske committed:
Move hook to inside form.

dshanske committed:
Token Endpoint will Refuse to Generate a Token if Scope is not Set

aaronpk committed:
move <form> markup higher to make it look a little better

dshanske committed:
Merge pull request #3 from aaronpk/token

minor UI improvements to IndieAuth UI

aaronpk committed:
standardize IndieAuth error handling

* ensures all HTTP status codes are returned appropriately
* returns OAuth errors using `error` and `error_description` instead of WordPress style errors
* adds more detailed error messages at various parts of the workflow to help debugging later

dshanske committed:
Merge pull request #4 from aaronpk/token

standardize IndieAuth error handling

dshanske committed:
Additional Error Fixes

pfefferle commented:
Sorry, but I can't get it to work...

pfefferle commented:
OK, it was a caching problem, seems to work now! Nice!

pfefferle commented:
I can use my Blog to login to Quill or to the IndieWeb Wiki, bit the login to WordPress does not work.

aaronpk commented:
It worked for me, but of course I had to set the authorization endpoint in the settings back to `https://indieauth.com/auth`.

If you have the authorization endpoint set to the built-in one, then it would be kind of stuck in a weird state. Which brings me to another point. How do I know what the URL for the built-in authorization endpoint is? Right now if I change it to indieauth.com, then I have no way of knowing what to change it to to make it go back. Blanking out the value just stores a blank value which breaks it. Can you make it revert to the default built-in URL if a blank value is saved?

aaronpk committed:
fix whitespace in form elements

with the `<?php` tags on the next line, it ends up including a bunch of whitespace in the values that are saved

pfefferle commented:
@dshanske I used my main blog (notiz.blog) to login to (pfefferle.org) and I stuck at a white page.

aaronpk commented:
@pfefferle what authorization server is it set to use in that case?

pfefferle commented:
@aaronpk the internal on, that @dshanske implemented...

aaronpk commented:
Yeah I don't think that makes any sense, since you'd have to be logged in before you can log in. It only makes sense to sign in to wordpress using your domain if you're using an external authorization server.

aaronpk commented:
I'm not really sure how to solve the UI around this. tbh I think it would make sense to pull out the "log in to your wordpress" feature into a separate plugin, preferably called RelMeAuth or IndieLogin (to match the upcoming indielogin.com that I'm working on), since logging in to wordpress with your domain isn't actually doing IndieAuth.

I'm worried that having both in the same plugin is going to continue to confuse people.

dshanske commented:
I can get behind a fork

dshanske commented:
The alternative is to have a button on the login form that links to a relmeauth form. I can code that.

aaronpk commented:
Not sure what you mean, how does that change anything?

pfefferle commented:
@aaronpk but why? I want to use may domain to login to different sites using the internal indieauth?

dshanske commented:
For the default setting, I think I should change the interface to a radio button for default.

dshanske commented:
@pfefferle You should be able to do that

dshanske commented:
I can go back and redesign the interface. But should it be merged first and a second PR?

pfefferle commented:
@aaronpk sure, I understand the difference between rel-me auth and indieauth and I see, that it does not make sense to login to notiz.blog using notiz.blog... but using notiz.blog to login to pfefferle.org should work and that isn't working at the moment...

aaronpk commented:
I don't see this IndieAuth plugin installed on pfefferle.org: https://pfefferle.org/wp-login.php There's only an OpenID option there. Can you clarify exactly what you are expecting to happen?

pfefferle commented:
I removed it after some failed tests...

aaronpk commented:
So to clarify, you installed this plugin on both sites, and you were trying to log in to one as the other? It would be nice to get more information about the error you're having so we can try to reproduce it.

pfefferle commented:
@aaronpk what I tried to do:

* I added `https://notiz.blog/` to the login field on pfefferle.org
* Browser redirects me to indieauth.com
* I clicked at the `IndieAuth` authentication
* Browser redirects me to notiz.blog with the authentication screen
* I accepted the authentication dialog
* Browser redirects me to `https://pfefferle.org/wp-login.php?redirect_to=https%3A%2F%2Fpfefferle.org%2Fwp-admin%2F&code=XXX` and stays blank.

I will investigate why it stops rendering and try to debug the code. Perhaps I can send a merge request later today.

aaronpk commented:
Which site was this plugin installed at? Both? And which site was configured to point to indieauth.com?

Technically we shouldn't need to end up on indieauth.com for this use case, though that might need to be a separate set of changes to the plugin.

pfefferle commented:
@aaronpk both sites have the plugin installed and both are set to the native endpoint.

pfefferle commented:
@aaronpk @dshanske I think I found one problem: https://github.com/indieweb/wordpress-indieauth/pull/20/files#diff-839da1f39b21c2dad138c57211f89cbdL129 uses an old options entry that is not used any more!

pfefferle commented:
I think it was hardcoded to indieauth.com in an old version and we have to replace it with the auto-dicovered authorisation endpoint of the used URL!

dshanske commented:
@pfefferle Tell me if this is the problem. You try to log into notiz.blog with pfefferle.org and you click both times, login with my domain... So, the first time, it redirects you to pfefferle.org and if you don't click login with your domain the second time it will show the authenticate page. By clicking the check box, you are telling pfefferle.org to log you in using indieauth...Creating a loop... I think that interface needs improvement.

pfefferle commented:
This is what I am trying...

![apr -06-2018 19-00-25](https://user-images.githubusercontent.com/2373/38433950-fcf8ad08-39cc-11e8-9026-27b705c52e86.gif)

aaronpk commented:
indieauth.com should never have been involved in that flow if both wordpress plugins are set to use the internal authorization endpoint. I think your idea that an old options entry is being used is right.

pfefferle commented:
Is the micropub plugin adding indieauth headers???

pfefferle commented:
@dshanske @aaronpk the micropub plugin adds the default header below the indieauth headers, so the discovery gets the false headers... Can we disable that in the micropub plugin?

After deactivation the micropub plugin it does not load indieauth.com but I still can not login, because I stuck at a white page at the end!

pfefferle commented:
I always get:

```
WP_OAuth_Response Object
(
    [links:protected] => Array
        (
        )

    [matched_route:protected] =>
    [matched_handler:protected] =>
    [data] => Array
        (
            [error] => invalid_grant
            [error_description] => Invalid authorization code
        )

    [headers] =>
    [status] => 400
)
```

pfefferle commented:
Btw. the Quill login works as expected...

dshanske commented:
The development version doesn't do that. Expect @snarfed to be pushing it soon to fix the post status bug.

dshanske commented:
I am going to merge it then work on improvements in a new PR.

dshanske committed:
Merge pull request #5 from aaronpk/token

fix whitespace in form elements

pfefferle commented:
@dshanske why have you merged it? I still stuck at a white page! and the error handling does not really work well!

dshanske commented:
I am going to work on a new PR now to address the issues. Apologies for the confusion, I thought you said it was working. I misread.

dshanske commented:
I think I need to fix the form and the error handling.

dshanske commented:
@pfefferle Will have the new PR tonight
